src/utils/parseMacros.ts

```diff
@@ -0,0 +1,90 @@
+import { LintConfig } from '../types/LintConfig'
+import { LineEndings } from '../types/LineEndings'
+import { trimComments } from './trimComments'
+
+interface Macro {
+  name: string
+  startLineNumber: number | null
+  endLineNumber: number | null
+  declaration: string
+  termination: string
+  parentMacro: string
+  hasMacroNameInMend: boolean
+  hasParentheses: boolean
+  mismatchedMendMacroName: string
+}
+
+export const parseMacros = (text: string, config?: LintConfig): Macro[] => {
+  const lineEnding = config?.lineEndings === LineEndings.CRLF ? '\r\n' : '\n'
+  const lines: string[] = text ? text.split(lineEnding) : []
+  const macros: Macro[] = []
+
+  let isCommentStarted = false
+  let macroStack: Macro[] = []
+  lines.forEach((line, index) => {
+    const { statement: trimmedLine, commentStarted } = trimComments(
+      line,
+      isCommentStarted
+    )
+    isCommentStarted = commentStarted
+    const statements: string[] = trimmedLine ? trimmedLine.split(';') : []
+
+    statements.forEach((statement) => {
+      const { statement: trimmedStatement, commentStarted } = trimComments(
+        statement,
+        isCommentStarted
+      )
+      isCommentStarted = commentStarted
+
+      if (trimmedStatement.startsWith('%macro')) {
+        const startLineNumber = index + 1
+        const name = trimmedStatement
+          .slice(7, trimmedStatement.length)
+          .trim()
+          .split('(')[0]
+        macroStack.push({
+          name,
+          startLineNumber,
+          endLineNumber: null,
+          parentMacro: macroStack.length
+            ? macroStack[macroStack.length - 1].name
+            : '',
+          hasParentheses: trimmedStatement.endsWith('()'),
+          hasMacroNameInMend: false,
+          mismatchedMendMacroName: '',
+          declaration: line,
```

Comment:
Need to use `statement` instead of `line` here.

For example(multiple statements in single line):
`  %macro somemacro();    %put &sysmacroname;  %mend somemacro;`